hazelcast/src/main/java/com/hazelcast/internal/cluster/impl/MembershipManager.java

```diff
@@ -0,0 +1,895 @@
+                    try {
+                        MembersView membersView = future.get();
+                        if (membersView.isLaterThan(latestMembersView)) {
+                            logger.fine("A more recent " + membersView + " is received from " + address);
+                            latestMembersView = membersView;
+
+                            // If we discover a new member via a fetched member list, we should also ask for its members view.
+                            // there are some new members added to the futures map. lets wait for their results.
+                            done &= !fetchMembersViewFromNewMembers(membersView, futures);
+                        }
+                    } catch (InterruptedException ignored) {
+                        Thread.currentThread().interrupt();
+                    } catch (ExecutionException ignored) {
+                        // we couldn't learn MembersView of 'address'. It will be removed from the cluster.
+                        EmptyStatement.ignore(ignored);
+                    }
+                } else if (!isMemberSuspected(address) && latestMembersView.containsAddress(address)) {
+                    // we don't suspect from 'address' and we need to learn its response
+                    done = false;
+                }
+            }
+
+            if (done) {
+                break;
+            }
+
+            try {
+                Thread.sleep(FETCH_MEMBERLIST_SLEEP_INTERVAL);
+            } catch (InterruptedException ignored) {
+                Thread.currentThread().interrupt();
+            }
+        }
+
+        return latestMembersView;
+    }
+
+    private boolean fetchMembersViewFromNewMembers(MembersView membersView, Map<Address, Future<MembersView>> futures) {
+        boolean isNewMemberPresent = false;
+
+        for (MemberInfo memberInfo : membersView.getMembers()) {
+            Address memberAddress = memberInfo.getAddress();
+            if (!(node.getThisAddress().equals(memberAddress)
+                    || isMemberSuspected(memberAddress)
+                    || futures.containsKey(memberAddress))) {
+                // this is a new member for us. lets ask its members view
+                logger.fine("Asking MembersView of " + memberAddress);
+                futures.put(memberAddress, invokeFetchMembersViewOp(memberAddress, memberInfo.getUuid()));
+                isNewMemberPresent = true;
+            }
+        }
+
+        return isNewMemberPresent;
+    }
+
+    private Future<MembersView> invokeFetchMembersViewOp(Address target, String targetUuid) {
+        Operation op = new FetchMembersViewOp(targetUuid).setCallerUuid(node.getThisUuid());
+
+        return nodeEngine.getOperationService()
+                .createInvocationBuilder(SERVICE_NAME, op, target)
+                .setTryCount(mastershipClaimTimeoutSeconds)
+                .setCallTimeout(TimeUnit.SECONDS.toMillis(mastershipClaimTimeoutSeconds)).invoke();
```

Comment:
Yes, it is a trade-off between availability and correctness. If it is shorter, the cluster will become available earlier on master failure but it can kick out some slow members during the process.